### DISCUSSION TITLE
Question about how to use MicrostructurePy for input numpy arrays

### DISCUSSION THREAD

**Discussion by ahmad681** [kip-hart/MicroStructPy#86] [General]
Hello, 
first thanks for this wonderful package, it is fantastic. It seems to be potential to serving my purpose for research. 
So I have two-phase micostructure generated. It is basically a 2D numpy array of 1's and 2's. The examples I studied so far using MicroPy seems to be self-generating their own microstructure and then mesh them. I would like to input my randomly generated two-phase microstructure using my numpy 2D array as input, and mesh it using MicrostructurePy. How do you advise in proceeding forward with it? I see an example of importing an image, but again I am trying to import a 2D array. 
Thank you a lot in advance. It looks something like the image below, it is nothing more than cartesian array of ones and twos. Many thanks in advance. 
![image](https://github.com/kip-hart/MicroStructPy/assets/77033621/c19a0ea3-b864-4fb1-b1ad-e0864832f9db)

**Comment by kip-hart** [comment]
Hi @ahmad681, thank you for reaching out! The from_image example file converts the image to a 2D numpy array in the beginning, so you should be able to reuse the remainder of the code.

In this block, the image is converted to a 2D array of 0s and 1s. https://github.com/kip-hart/MicroStructPy/blob/master/src/microstructpy/examples/from_image.py#L20

Feel free to use the code in this file. If you define bin_nums to be equal to your 2D array, the rest of the code should work. Let me know if I can help further.

**Reply by ahmad681** [reply]
This is awesome, thanks a lot @kip-hart
I am assuming similar story goes with reading 3D arrays as well, correct?

**Reply by kip-hart** [reply]
So for 3D I'd probably use the RasterMesh class https://docs.microstructpy.org/en/latest/api/meshing/rastermesh.html

The class works for 2D as well. I hadn't written that at the time I made the examples. I should probably go back and update the example. Let me know if you have any questions.

**Reply by kip-hart** [reply]
Hang on, do you want the 3D case to generate an unstructured grid from a 3D raster?

**Reply by ahmad681** [reply]
Eventually, I will generate a 3D array in numpy also again of two-phase structure. The array is in xyz, currently I am just going simple and testing it in a 2D grid xy. 
But ultimately by the end of this week I will go for 3D, so I will input my 3D array into microstructurePy and have it generate a mesh for me, sorry if that didn't make sense earlier. So I guess yes, you are correct, generate 3D mesh from a 3D arrays of ones and twos.

**Reply by kip-hart** [reply]
Ok yes I understand now. RasterMesh won't be useful in this case.

You can start with the 2D PolyMesh code in the from_images example when going to 3D. Things might get a little more complicated, such as defining the facets and neighbor pairs between voxels. The general principle is the same tho, you're essentially building all the connectivity for the PolyMesh.

Let me know if you have any other questions.

**Reply by ahmad681** [reply]
I see, sounds like a plan. I will keep this post updated even if I have it working, I will share the code here.
Thanks for the help Hart :)
